File: serial_comm/SerialProtocol.cpp
#include "SerialProtocol.h"

#include "Messages.h"
#include "HandshakeMessage.h"
#include "HeartbeatMessage.h"
#include "Arduino.h"

Vector<unsigned char> SerialProtocol::encodeMessage(const MessageBase& message) const
{
  Vector<unsigned char> data = Vector<unsigned char>(1u);
  data.push_back((unsigned char)message.getType());

  Vector<unsigned char> payload = message.encode();

  for (unsigned int i = 0; i < payload.size(); ++i)
  {
    data.push_back(payload.element_at(i));
  }

  return data;
}

MessageBase* SerialProtocol::decodeMessage(const unsigned char* data, const uint16_t size) const
{
  Messages messageId = (Messages)data[0];
  
  MessageBase* retVal = NULL;

  switch (messageId)
  {
    case Messages::HANDSHAKE:
    {
      retVal = new HandshakeMessage();
    }
    break;
    case Messages::HEARTBEAT:
    {
      retVal = new HeartbeatMessage();
    }
    break;
    default:
    break;
  }

  if (NULL != retVal)
  {
    retVal->decode(&data[1], (size - 1u));
  }

  return retVal;
}

File: python/arduino_termpoint/comm_messaging/comm_serial/serial_protocol.py
from ..protocol import Protocol
from .messages import Messages
from .handshake_message import HandshakeMessage
from .heartbeat_message import HeartbeatMessage
import struct

class SerialProtocol(Protocol):
    def encode(self, message):
        data = struct.pack('<B', message.get_msg_type())
        payload = message.encode()

        if None != payload:
            data += payload

        return data

    def decode(self, data):
        msg_type = struct.unpack('<B', data[0:1])[0]
        msg = None
        if Messages.HANDSHAKE == msg_type:
            msg = HandshakeMessage()
        elif Messages.LOG == msg_type:
            print("Log received but not handled")
        elif Messages.HEARTBEAT == msg_type:
            msg = HeartbeatMessage()

        if None != msg:
            msg.decode(data[1:])

        return msg